# Wstęp do Sztucznej Inteligencji - rok akademicki 2021/2022

Przed rozpoczęciem pracy z notatnikiem zmień jego nazwę zgodnie z wzorem: `NrAlbumu_Nazwisko_Imie_PoprzedniaNazwa`.

Przed wysłaniem notatnika upewnij się, że rozwiązałeś wszystkie zadania/ćwiczenia.

# Temat: Wnioskowanie oparte na wiedzy niepewnej. Wnioskowanie bayesowskie
Zapoznaj się z treścią niniejszego notatnika czytając i wykonując go komórka po komórce. Wykonaj napotkane zadania/ćwiczenia.

## Zadanie 1  (obowiązkowe, 5pkt.)
Zapożyczone z ćwiczeń do wykładu [Arabas, Cichosz](http://wazniak.mimuw.edu.pl/index.php?title=Sztuczna_inteligencja/SI_%C4%86wiczenia_4)

### Zamodeluj i odpowiedz na pytania.

W śledztwie dotyczącym zabójstwa inspektor Bayes rozważa dwie hipotezy:

- $H_1$ główny podejrzany zabił,
- $H_2$ główny podejrzany nie zabił, 

oraz następujące możliwe fakty:

- $E_1$ na miejscu zbrodni znaleziono odciski palców głównego podejrzanego,
- $E_2$ główny podejrzany nie ma alibi na czas popełnienia zabójstwa,
- $E_3$ główny podejrzany miał motyw zabicia ofiary,
- $E_4$ główny podejrzany był widziany w sądziedztwie miejsca, w którym mieszka nielegalny handlarz bronią,
- $E_5$ świadek zbrodni podał rysopis zabójcy nie pasujący do głównego podejrzanego. 

Zależności między takimi faktami a hipotezami opisują następujące prawdopodobieństwa:

$P(E_1|H_1)=0.7,\qquad P(E_1|H_2)=0.3,$

$P(E_2|H_1)=0.8,\qquad P(E_2|H_2)=0.4,$

$P(E_3|H_1)=0.9,\qquad P(E_3|H_2)=0.5,$

$P(E_4|H_1)=0.4,\qquad P(E_4|H_2)=0.2,$

$P(E_5|H_1)=0.2,\qquad P(E_5|H_2)=0.4.$ 

__W którym przypadku prawdopodobieństwo popełnienia zabójstwa byłoby największe?__

1. Gdyby znaleziono na miejscu zbrodni jego odciski palców.
2. Gdyby stwierdzono, że nie miał alibi i miał motyw.
3. Gdyby znaleziono na miejscu zbrodni jego odciski palców oraz stwierdzono, że był widziany w sąsiedztwie miejsca, w którym mieszka nielegalny handlarz bronią, ale świadek zbrodni podał rysopis zabójcy nie pasujący do głównego podejrzanego.

### TWÓJ PROGRAM:

In [ ]:
p_H1 = 0.5
p_H2 = 0.5
p_E1_H1 = 0.7
p_E2_H1 = 0.8
p_E3_H1 = 0.9
p_E4_H1 = 0.4
p_E5_H1 = 0.2

p_E1_H2 = 0.3
p_E2_H2 = 0.4
p_E3_H2 = 0.5
p_E4_H2 = 0.2
p_E5_H2 = 0.4

p_E1 = p_H1*p_E1_H1+p_H2*p_E1_H2
p_E2 = p_H1*p_E2_H1+p_H2*p_E2_H2
p_E3 = p_H1*p_E3_H1+p_H2*p_E3_H2
p_E4 = p_H1*p_E4_H1+p_H2*p_E4_H2
p_E5 = p_H1*p_E5_H1+p_H2*p_E5_H2

p_H1_E1 = (p_H1 * p_E1_H1) / p_E1
print("1. ", p_H1_E1)

p_H1_E2_E3 = (p_E2_H1 * p_E3_H1 * p_H1) / (p_E2_H1 * p_E3_H1 * p_H1 + p_E2_H2 * p_E3_H2 * p_H2)
print("2. ", p_H1_E2_E3)

p_H1_E1_E4_E5 = (p_E1_H1 * p_E4_H1 * p_E5_H1 * p_H1) / (p_E1_H1 * p_E4_H1 * p_E5_H1 * p_H1 + p_E1_H2 * p_E4_H2 * p_E5_H2 * p_H2)
print("3. ", p_H1_E1_E4_E5)

1.  0.7
2.  0.7826086956521738
3.  0.7000000000000001


### ODPOWIEDŹ:

2. Gdyby stwierdzono, że nie miał alibi i miał motyw.

## Zadanie 2  (obowiązkowe, 5pkt.)

### Zamodeluj i odpowiedz na pytania.
System alarmowy w mieszkaniu, reaguje na włamania oraz, niestety, również na drobne trzęsienia (ziemi). Sąsiedzi John i Mary są umówieni, żeby zadzwonić do właściciela gdy usłyszą alarm. John jest nadgorliwy i bierze różne zdarzenia (np. dzwonek telefonu) za sygnał alarmowy (i wtedy zawsze dzwoni). Mary rozpoznaje alarm poprawnie, lecz często słucha głośnej muzyki i może go w ogóle nie usłyszeć. 

Sieć przekonań dla systemu alarmowego wygląda następująco:
![bsiec.PNG](http://torus.uck.pk.edu.pl/~amarsz/images/bsiec.PNG)

__Jakie jest prawdopodobieństwo, że:__
1. włączy się alarm?
2. doszło do włamanie jeśli wiadom, że włączył się alarm?
3. zdarzyło się trzęsienie ziemi jeśli wiadomo, żę włączył się alarm?
1. w razie włamania ktoś zadzwoni?
2. zawiadomienie o włamaniu jest fałszywe?
3. rozległ się alarm, przy czym nie wystąpiło ani trzęsienie ziemi ani włamanie, ale oboje John i Mary zadzwonili? (prawd. bezwarunkowe)

TWÓJ PROGRAM:

In [2]:
import pymc3 as pm
import arviz as az
import numpy as np

with pm.Model() as model:
    earthquake = pm.Bernoulli('earthquake', 0.002)
    burglary = pm.Bernoulli('burglary', 0.001)
    alarm_p = pm.Deterministic('alarm_p', pm.math.switch(earthquake,pm.math.switch(burglary, 0.95, 0.29), pm.math.switch(burglary, 0.94, 0.001)))
    alarm = pm.Bernoulli('alarm', alarm_p)
    johnCalls_p = pm.Deterministic('johnCalls_p', pm.math.switch(alarm, 0.9, 0.05))
    johnCalls = pm.Bernoulli('johnCalls', johnCalls_p)
    maryCalls_p = pm.Deterministic('maryCalls_p', pm.math.switch(alarm, 0.7, 0.01))
    maryCalls = pm.Bernoulli('maryCalls', maryCalls_p)

    trace = pm.sample(100000, chains=1, return_inferencedata=True)


p_alarm_on = trace.posterior['alarm'].values[0].sum()/len(trace.posterior['alarm'].values[0])
print("1. ", p_alarm_on)

p_burglary_alarm_on = (trace.posterior['burglary'].values[0]*trace.posterior['alarm'].values[0]).sum()/trace.posterior['alarm'].values[0].sum()
print("2. ", p_burglary_alarm_on)

p_earthquake_alarm_on = (trace.posterior['earthquake'].values[0]*trace.posterior['alarm'].values[0]).sum()/trace.posterior['alarm'].values[0].sum()
print("3. ", p_earthquake_alarm_on)

p_burglary_true_call = (trace.posterior['burglary'].values[0] * trace.posterior['johnCalls'].values[0] * trace.posterior['maryCalls'].values[0]).sum() / (trace.posterior['johnCalls'].values[0] * trace.posterior['maryCalls'].values[0]).sum()
print("4. ", p_burglary_true_call)

p_burglary_false_call = np.logical_not(trace.posterior['burglary'].values[0]).sum()/len(np.logical_not(trace.posterior['burglary'].values[0]))
print("5. ", p_burglary_false_call)

p_alarm_and_both_call = (trace.posterior['alarm'].values[0] * trace.posterior['johnCalls'].values[0] * trace.posterior['maryCalls'].values[0]).sum() / len(trace.posterior['alarm'].values[0])
print("6. ",p_alarm_and_both_call)

Sequential sampling (1 chains in 1 job)
BinaryGibbsMetropolis: [earthquake, burglary, alarm, johnCalls, maryCalls]


Sampling 1 chain for 1_000 tune and 100_000 draw iterations (1_000 + 100_000 draws total) took 53 seconds.
Only one chain was sampled, this makes it impossible to run some convergence checks


1.  0.00287
2.  0.3310104529616725
3.  0.26480836236933797
4.  0.25301204819277107
5.  0.99895
6.  0.00194


### ODPOWIEDŹ:

1.  0.00287
2.  0.3310104529616725
3.  0.26480836236933797
4.  0.25301204819277107
5.  0.99895
6.  0.00194 #tego nie jestm pewnien czy jest dobrze

__UWAGA:__ Zwróć uwagę na wielkości podanych prawdopodobieńst aby dobarć odpowiednią liczbę symulacji.

&copy; Katedra Informatyki, Politechnika Krakowska